catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+            throw new SubscriptionStoreException(
+                    "Duplicate factory registered. This is an API client error");
+        }
+
+        factories.put(factory.getTypeName(), factory);
+        LOGGER.debug("SubscriptionFactory registered for {}", factory.getTypeName());
+
+        StreamSupport.stream(subscriptionsCache.spliterator(), false)
+                .map(Cache.Entry::getValue)
+                .filter(CachedSubscription::isNotRegistered)
+                .filter(sub -> sub.isType(factory.getTypeName()))
+                .forEach(sub -> sub.registerSubscription(factory));
+    }
+
+    /**
+     * Called when the OSGi framework unregisters a factory.
+     *
+     * @param factory the {@link SubscriptionFactory} that was unregistered.
+     * @throws SubscriptionStoreException if an unexpected registration tried to occur.
+     */
+    public synchronized void unbindFactory(SubscriptionFactory factory) {
+        if (factory == null) {
+            LOGGER.debug("Subscription container unbinding was given a null factory. ");
+            return;
+        }
+
+        SubscriptionFactory removedFactory = factories.remove(factory.getTypeName());
+        if (removedFactory == null) {
+            throw new SubscriptionStoreException(format(
+                    "Binding synchronization is wrong. No factory [%s] exists to unbind. ",
+                    factory.getTypeName()));
+        }
+        LOGGER.debug("SubscriptionFactory removed for {}", factory.getTypeName());
+    }
+
+    /**
+     * Helper method for core insert logic.
+     */
+    private void doInsert(Subscription subscription, SubscriptionMetadata metadata) {
+        CachedSubscription cachedSubscription = createCachedSubscription(metadata);
+        cachedSubscription.registerSubscription(subscription);
+        try {
+            subscriptionsCache.put(metadata.getId(), cachedSubscription);
+        } catch (CacheException e) {
+            cachedSubscription.unregisterSubscription();
+            throw new SubscriptionStoreException("Problem writing to the cache. ", e);
+        }
+    }
+
+    /**
+     * Helper method for core delete logic.
+     */
+    private CachedSubscription doDelete(String subscriptionId) {
+        CachedSubscription cachedSubscription = subscriptionsCache.get(subscriptionId);
+        try {
+            subscriptionsCache.remove(subscriptionId);
+        } catch (CacheException e) {
+            throw new SubscriptionStoreException("Problem deleting from cache. ", e);
+        }
+        cachedSubscription.unregisterSubscription();
+        return cachedSubscription;
+    }
+
+    /**
+     * Ensure the identifier points to a valid subscription, and that it exists in the container.
+     *
+     * @param operation customize the message based on where this validation is occurring.
+     */
+    private void validateContainment(SubscriptionIdentifier identifier, String operation) {
+        if (get(identifier) == null) {
```

Comment:
That works. I was trying to avoid the negation, and declare conditions as they **are** instead of what they **aren't**.